### PR TITLE
API Client that must be handled by implementing gRPC API Server with business logic

### PR DESCRIPTION
**Pull request by LinMAD**
Some instructions about test assignment and bare Golang code for API Client that generates requests to the server.

Things to cover
- [x] Review Instructions if they clear
- [x] CI
- [x] Docker

### CODE REVIEW

File: .vale.ini [hunk 1]
```diff
@@ -0,0 +1,9 @@
+StylesPath = styles
+
+Packages = https://github.com/coopnorge/vale-coop/releases/latest/download/Coop.zip
+
+Vocab = General, CoopAbbreviations, CoopGeneral
```
Comment (nhhagen) [inline .vale.ini]: These are loaded automatically by the package

```suggestion
```

File: .vale.ini [hunk 2]
```diff
@@ -0,0 +1,9 @@
+StylesPath = styles
+
+Packages = https://github.com/coopnorge/vale-coop/releases/latest/download/Coop.zip
+
+Vocab = General, CoopAbbreviations, CoopGeneral
+
+[*.md]
+TokenIgnores = [\w][\w.]*@[\w*][\w.]*[\w]
+BasedOnStyles = Vale
```
Comment (nhhagen) [inline .vale.ini]: This is also activated by the loaded package.

```suggestion
```

File: api/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# Contribute instructions
+
+Make sure you have before:
+
+- [Go](https://go.dev/doc/install)
+- [Buf for proto-files management](https://buf.build/docs/installation)
+
+Check make file in root of the project for compiling go code from proto-files.
```
Comment (nhhagen) [inline api/README.md]:
```suggestion
Check the `Makefile` in root of the project for compiling go code from proto-files.
```

File: mkdocs.yml
Comment (nhhagen) [inline mkdocs.yml:1]: `mkdocs.yml` should not be committed and added to `.gitignore`

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    container_name: interview_backend_coop_logistics
+    environment:
+      - APP_PORT=50051
+    tty: true
+    ports:
+      - "50051:50051"
+  techdocs:
+    build:
+      context: .
+      dockerfile: internal/techdocs/Dockerfile
```
Comment (nhhagen) [inline docker-compose.yml]: Why are we not using our common pattern of having this in `/docker-compose/`?

Comment (LinMAD) [reply docker-compose.yml]: Minimized "Coop specifics" so for participants would be a bit easier, could create questions like "why docker-compose folder with docker images? etc"

We can move it to the root as well, not a big deal.

File: docs/INDEX.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# Interview Assignment - Coop Logistics API
+
+## Acceptance Criteria
+
+- All three parts must be done (One, Two, Three).
```
Comment (yurydavydov) [inline docs/INDEX.md]: You could delete the part in braces. It does not add any additional information.

File: docs/INDEX.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+- API must be sustainable to handle requests without dropping them.
+
+## Assignment Appendix
+
+Coop Logistics Engine, simulates several cargo units that
```
Comment (yurydavydov) [inline docs/INDEX.md]: "," after "Engine" is unnecessary.

File: docs/INDEX.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+deliver goods between different warehouses in the world.
+
+Coop Logistics Engine sends `gRPC` requests to `127.0.0.1:50051`,
+checkout service in
+[docker-compose](../docker-compose.yml) "interview_backend_coop_logistics"
+
+## Assignment Details
```
Comment (yurydavydov) [inline docs/INDEX.md]: Instead of "Assignment Details", I would call it "Assignment".

Further, I would rephrase:
"Develop an API system that matches the criteria described in parts. Here is [Swagger documentation with descriptions.](../api/logistics.swagger.json)"

to

"The result of this task should be an API system that fulfils the criteria described in three parts below. Here is the link to the API documentation [Swagger documentation with descriptions.](../api/logistics.swagger.json)"

File: docs/INDEX.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+Develop an API system that matches the criteria described in parts,
+here is [Swagger documentation with descriptions.](../api/logistics.swagger.json)
+
+Don't hesitate to ask questions for clarification if you have them.
```
Comment (yurydavydov) [inline docs/INDEX.md]: This "Call for action" and wishes good luck I would place at the end after part three.

File: docs/INDEX.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+*Good luck and we hope you will have fun!*
+___
+
+### Part One: Implement Backend API
+
+Your first task is to implement API server
```
Comment (yurydavydov) [inline docs/INDEX.md]: I would rephrase it as "Implement the API server for that..."

1. You do not need to refer to the candidate "Your...". It's clear from the context who has to work on the task.
2. It's already in part one, so you do not have to tell "first". Keep it short.

File: docs/INDEX.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+for that [proto-file](../api/v1/logistics.proto).
+
+The solution should output a log message to STDOUT once per second with:
+
+- Number of received messages / second.
```
Comment (yurydavydov) [inline docs/INDEX.md]: A list with just one item does not look good. Rephrase to "The solution should output a log message to STDOUT once per second with the number of received messages per second." Same in the part two.

File: docs/INDEX.md (hunk 7 above)
Comment (LinMAD) [reply docs/INDEX.md]: Nice, also added a link to protifle

File: docs/INDEX.md (hunk 8 above)
Comment (LinMAD) [reply docs/INDEX.md]: Maybe, but that feels like it will be a bit detached from the assignment task, where parts are pure instructions without distraction ;)

File: .github/workflows/build.yaml [hunk 11]
```diff
@@ -1,16 +1,39 @@
 on:
-  push: {}
+  push:
+    branches:
+      - main
   pull_request: {}
-concurrency:
-  group: ${{ github.workflow }}-${{ github.ref }}
-  cancel-in-progress: true
```
Comment (nhhagen) [inline .github/workflows/build.yaml]: We should keep this so that we cancel duplicate builds

File: api/v1/logistics.proto [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    int64 cargo_unit_id = 1;
+    // warehouse_id is unique id
+    int64 warehouse_id = 2;
+    // the message contains information about the announcement
+    string message = 3;
+}
+
+// Location where entity now located in X,Y Axis
+message Location {
+    int64 x = 1;
```
Comment (ezorrio) [inline api/v1/logistics.proto]: might be uint32/uint64, based on generator code

File: docs/INDEX.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
+# Interview Assignment - Coop Logistics API
```
Comment (ezorrio) [inline docs/INDEX.md:1]: As discussed, would be great to mention whether to implement client or server part, as we don't want to leave a room for different interpretation

File: api/v1/logistics.proto (hunk 12 above)
Comment (LinMAD) [reply api/v1/logistics.proto]: Yeah, random generator anyway using only absolute values

File: api/README.md [hunk 14]
```diff
@@ -0,0 +1,21 @@
+# Contribute instructions
+
+Make sure you have before:
```
Comment (stephendrake) [inline api/README.md]: Perhaps:
```suggestion
Prerequisites:
```

File: docs/INDEX.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+
+**Implement Server side part that can handle client requests
+with specific business logic.**
+
+## Acceptance Criteria
+
+- All three parts must be done.
+- API must be sustainable to handle requests without dropping them.
```
Comment (stephendrake) [inline docs/INDEX.md]:
```suggestion
- API must be capable of handling all requests sent by this client without dropping them.
```

File: docs/INDEX.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+It sends `gRPC` requests to `127.0.0.1:50051`,
+look at service in
+[docker-compose](../docker-compose.yml) "interview_backend_coop_logistics"
+
+## Assignment
+
+The result of this task should be an API system that fulfils the criteria
+described in three parts below. Here is the link to the API documentation
+[Swagger documentation](../api/logistics.swagger.json)
+or [Protofile](../api/v1/logistics.proto) with descriptions.
```
Comment (stephendrake) [inline docs/INDEX.md]: Minor: This line about the API specs is probably better suited in the "Part One" section, but as it's already mentioned there so one could probably remove it here.

Also, as the client expects a gRPC endpoint, so it's perhaps not worth mentioning the OpenAPI v2 spec until it's required.

File: docs/INDEX.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+Don't hesitate to ask questions for clarification if you have them.
+
+*Good luck and we hope you will have fun!*
+___
+
+### Part One: Implement Backend API
+
+Implement the API server for that
+[proto-file](../api/v1/logistics.proto).
```
Comment (stephendrake) [inline docs/INDEX.md]:
```suggestion
Implement an API sever that provides the following service over gRPC: 
[proto-file](../api/v1/logistics.proto).
```

File: api/v1/logistics.proto [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    int64 cargo_unit_id = 1;
+    // warehouse_id is unique id
+    int64 warehouse_id = 2;
+    // the message contains information about the announcement
+    string message = 3;
+}
+
+// Location where entity now located in X,Y Axis
+message Location {
+    uint32 x = 1;
```
Comment (stephendrake) [inline api/v1/logistics.proto]: Probably not that important as it's for an assessment, but does one not generally use latitude and longitude in logistics?

Comment (LinMAD) [reply api/v1/logistics.proto]: @stephendrake We can but for simplicity, it's just an imaginary X,Y-Axis chart for immersive experience we can rename it :D

File: docs/INDEX.md (hunk 16 above)
Comment (LinMAD) [reply docs/INDEX.md]: Yeah let's do it like that, when we will require to implement it as HTTP server like REST then we can return it back. Will refactor it to be more generic, like here is a folder with API specs.

File: api/v1/logistics.proto (hunk 18 above)
Comment (stephendrake) [reply api/v1/logistics.proto]: Up to you, but if anything then just to avoid the inevitable "why no lat/long" that I would ask if I was doing the interview.

File: docs/INDEX.md [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+### Part Two: Store Delivery Units
+
+The cargo units run between the different warehouses in the world.
+
+Your task is to store these delivery paths,
+data comes into the API you implemented in *Part One*.
+Write at least one unit test that verifies that your solution works.
+
+The solution should output a log message to STDOUT once per second with:
```
Comment (stephendrake) [inline docs/INDEX.md]:
```suggestion
```
Perhaps just rework this bit for clarity. Either this or the line below and subsequent once-per-second requirements are are duplicated.

File: README.md [hunk 20]
```diff
@@ -1,41 +1,23 @@
-# github-template-default
+# Interview Assignment - Coop Logistics API
 
-This is a basic template for when creating a Github repository.
+The test assignment for a Software Engineer
+-> [instructions](docs/INDEX.md).
 
-## Initialization instructions:
+Example of output:
```
Comment (stephendrake) [inline README.md]:
```suggestion
Example of client program output:
```

**Review by stephendrake — COMMENTED** [review]
A couple points I missed before.

File: docs/INDEX.md (hunk 8 above)
Comment (LinMAD) [reply docs/INDEX.md]: Updated public part of that, internal code was not changed :)

**Comment by LinMAD** [comment]
@stephendrake Just a few bits that I added:
- Config support
- Updated docker-compse - now it will boost up setup since there is now a solution folder and docker-compse.yml preconfigured
- Minor ASCII printer now aligns rows length to values (pretty print)

File: docker-compose.yml [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    networks:
+      - coop-norge-interview-network
+  interview_backend_coop_logistics_client:
+    build:
+      context: .
+      dockerfile: docker-compose/Dockerfile-Project
+    container_name: interview_backend_coop_logistics_client
+    environment:
+      - CLIENT_SERVICE_PORT=50051
+      - CLIENT_SERVICE_HOST=interview_backend_coop_logistics_server
```
Comment (LinMAD) [inline docker-compose.yml:26]: Yeah maybe I rush it a bit, let me know if the naming is confusing :D

**Review by ezorrio — APPROVED** [review]

**Review by stephendrake — APPROVED** [review]

**Comment by LinMAD** [comment]
@nhhagen Any other contributions or are we good to go?

**Comment by nhhagen** [comment]
Looks good, I focused on the case description for the candidate. Do we have people to test this?

**Review by nhhagen — APPROVED** [review]

**Comment by LinMAD** [comment]
> Looks good, I focused on the case description for the candidate. Do we have people to test this?

We can speak about that in another thread :)
